src/read/gnu_compression.rs

```diff
@@ -0,0 +1,49 @@
+use crate::read::{self, Error, ReadError as _};
+use crate::{endian, CompressedFileRange, CompressionFormat, ReadRef as _, U32Bytes};
+
+// Attempt to parse the the CompressedFileRange for a section using the GNU-style
+// inline compression header format. This is used by the Go compiler in Mach-O files
+// as well as by the GNU linker in some ELF files. This function will return Ok(None)
+// if the section does not have the specified name prefix.
+pub(super) fn maybe_compressed_file_range<'data>(
+    section: &impl read::ObjectSection<'data>,
+    name_prefix: &str,
+) -> read::Result<Option<CompressedFileRange>> {
+    let name = match section.name() {
+        Ok(name) => name,
+        // I think it's ok to ignore this error?
+        Err(_) => return Ok(None),
+    };
+    if !name.starts_with(name_prefix) {
+        return Ok(None);
+    }
+    let (_section_offset, section_size) = section
+        .file_range()
+        .read_error("Invalid GNU compressed section type")?;
+    let mut offset = 0;
+    let data = section.data()?;
+    // Assume ZLIB-style uncompressed data is no more than 4GB to avoid accidentally
+    // huge allocations. This also reduces the chance of accidentally matching on a
+    // .debug_str that happens to start with "ZLIB".
+    if data
+        .read_bytes(&mut offset, 8)
+        .read_error("GNU compressed section is too short")?
+        != b"ZLIB\0\0\0\0"
+    {
+        return Err(Error("Invalid ELF GNU compressed section header"));
+    }
+    let uncompressed_size = data
+        .read::<U32Bytes<_>>(&mut offset)
+        .read_error("GNU compressed section is too short")?
+        .get(endian::BigEndian)
+        .into();
+    let compressed_size = section_size
+        .checked_sub(offset)
+        .read_error("GNU compressed section is too short")?;
+    Ok(Some(CompressedFileRange {
+        format: CompressionFormat::Zlib,
+        offset,
```

Comment:
This offset is wrong. Fix this by passing `self.file.data` into this function and use it with `section_offset` instead of using `section.data()`.

Also possibly see if you can avoid passing in `section` at all.

Reply:
Done, I moved the section name processing back up to the caller and got rid of the whole "maybe" aspect. See if it pleases.